src/cargo/core/compiler/job_queue.rs

```diff
@@ -827,20 +844,21 @@ impl<'cfg> DrainState<'cfg> {
             self.note_working_on(cx.bcx.config, unit, fresh)?;
         }
 
-        let doit = move || {
-            let state = JobState {
-                id,
-                messages: messages.clone(),
-                rmeta_required: Cell::new(rmeta_required),
-                _marker: marker::PhantomData,
-            };
-
-            let mut sender = FinishOnDrop {
-                messages: &messages,
-                id,
-                result: Err(format_err!("worker panicked")),
+        let doit = move |state| {
+            // Use `catch_unwind` to guarantee that a `Finish` message is sent even if our job
+            // panics. We shouldn't panic unless there's a bug in Cargo, so we just need to make
+            // sure nothing hangs by accident.
+            let result = match panic::catch_unwind(panic::AssertUnwindSafe(|| job.run(&state))) {
```

Comment:
Can this be done using a destructor like before instead of `catch_unwind`? In theory Cargo has no need to catch panics and works with `-Zpanic=abort`, and I think a destructor should work here as well?

Reply:
I found this to be cleaner. I can revert 69bc1c3e305f9d07611f554cf00f7075e4e6a21d if you prefer though.

Reply:
Yeah I agree this is a bit cleaner, but for me the functionality of "don't rely on `catch_unwind` unless you really need to" is more important.

Reply:
Ok, done.